vision/api/Detect/Program.cs

```diff
@@ -318,6 +333,109 @@ private static object DetectDocText(Image image)
             return 0;
         }
 
+        // [START vision_async_detect_document_ocr]
+        private static object DetectDocument(string gcsSourceUri, string gcsDestinationUri)
+        {
+
+            // Supported mime_types are: 'application/pdf' and 'image/tiff'
+            string mimeType = "application/pdf";
+
+            // How many pages should be grouped into each json output file.
+            // With a file of 5 pages
+            int batchSize = 2;
+
+            var client = ImageAnnotatorClient.Create();
+
+            var feature = new Feature{

```

Comment:
I found these little variables very confusing when I read the sample.

I have to mentally process [FORTY THREE LOC](https://github.com/telpirion/dotnet-docs-samples/blob/2dd06fa770c4688370ae62c5f7de9cedf83b5fa3/vision/api/Detect/Program.cs#L340-L383) ([43 LOC](https://github.com/telpirion/dotnet-docs-samples/blob/2dd06fa770c4688370ae62c5f7de9cedf83b5fa3/vision/api/Detect/Program.cs#L340-L383)) _before the API is even called._

**AND...** the sample is _backwards_. The tiniest structure (feature) is created _first_, but it's one of the _last_ parts of the request to be constructed.

The only way I could comprehend this sample was to scroll down to find the `client.AsyncBatchAnnotateFiles(requests)` and then read UPWARDS to put the puzzle pieces together

----

Oof! – So. What do we do? How can we help make this easier to comprehend?

----

Well. As a developer, I would like to start with the API call. If that's not possible, let's at least start by building up the **Request** <– _an object which a developer would _expect_ to find in a client library/API sample_

Think about reading the API Reference. You might look at an object like [AnnotateImageRequest](https://cloud.google.com/vision/docs/reference/rpc/google.cloud.vision.v1#google.cloud.vision.v1.AnnotateImageRequest) & your goal is to understand the structure of the request, right? If you look at the request object from the OUTSIDE –> Inwards things make more sense

Compare:

 - MimeType
 - BatchSize
 - Feature
 - GcsSource
 - InputConfig
 - GcsDestination
 - OutputConfig
 - Async Annotate File Request
 - Operation!

To this:

 - Async Annotate File Request
   - Input Config
    - GcsSource
    - MimeType
  - OutputConfig
    - GcsDestination
    - BatchSize
  - Features
 - Operation!

One is **flat**. The other is **hierarchical**.

This is part of why I always recommend constructing the request object _inline_, when kicking off the API method, eg:

```cs
var client = LanguageServiceClient.Create();
var response = client.AnalyzeEntitySentiment(new Document()
{
    Content = text,
    Type = Document.Types.Type.PlainText
});
```

At the end of the day, it's about whatever we think is the (a) easiest to consume (b) highlights what we want to highlight (c) does not include extraneous code (d) is not overly long, pay attention to vertical space consumption, aka LOC, as well as horizontal usage/display

----

Here's an IDEA for how I might have coded this API request (note: I haven't gotten to the GCS JSON bottom half yet)

```cs
var client = ImageAnnotatorClient.Create();

var asyncRequest = new AsyncAnnotateFileRequest
{
    InputConfig = new InputConfig
    {
        GcsSource = new GcsSource { Uri = gcsSourceUri },
        // Supported mime_types are: "application/pdf" and "image/tiff"
        MimeType = "application/pdf"
    },
    OutputConfig = new OutputConfig
    {
        GcsDestination = new GcsDestination { Uri = gcsDestinationUri },
        // How many PDF/TIFF pages should be grouped into each JSON file
        BatchSize = 2,
    }
};

asyncRequest.Features.Add(new Feature {
    Type = Feature.Types.Type.DocumentTextDetection
});

var operation = client.AsyncBatchAnnotateFiles(requests: new[]{ asyncRequest });
```

### My 2¢